src/test/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtensionTest.java

```diff
@@ -841,4 +841,289 @@
+
+            + "    public MoshiJsonAdapter(Moshi moshi, Type[] types) {\n"
+            + "      this.itemAdapter = moshi.adapter(types[0]);\n"
+            + "      this.nameAdapter = moshi.adapter(String.class);\n"
+            + "    }\n"
+            + "  \n"
+            + "    @Override public Foo<V> fromJson(JsonReader reader) throws IOException {\n"
+            + "      reader.beginObject();\n"
+            + "      V item = null;\n"
+            + "      String name = null;\n"
+            + "      while (reader.hasNext()) {\n"
+            + "        String name_ = reader.nextName();\n"
+            + "        if (reader.peek() == JsonReader.Token.NULL) {\n"
+            + "          reader.skipValue();\n"
+            + "          continue;\n"
+            + "        }\n"
+            + "        switch (name_) {\n"
+            + "          case \"item\": {\n"
+            + "            item = itemAdapter.fromJson(reader);\n"
+            + "            break;\n"
+            + "          }\n"
+            + "          case \"name\": {\n"
+            + "            name = nameAdapter.fromJson(reader);\n"
+            + "            break;\n"
+            + "          }\n"
+            + "          default: {\n"
+            + "            reader.skipValue();\n"
+            + "          }\n"
+            + "        }\n"
+            + "      }\n"
+            + "      reader.endObject();\n"
+            + "      return new AutoValue_Foo(item, name);\n"
+            + "    }\n"
+            + "  \n"
+            + "    @Override public void toJson(JsonWriter writer, Foo<V> value) throws IOException {\n"
+            + "      writer.beginObject();\n"
+            + "      writer.name(\"item\");\n"
+            + "      itemAdapter.toJson(writer, value.item());\n"
+            + "      writer.name(\"name\");\n"
+            + "      nameAdapter.toJson(writer, value.name());\n"
+            + "      writer.endObject();\n"
+            + "    }\n"
+            + "  }\n"
+            + "}"
+    );
+
+    assertAbout(javaSource())
+            .that(source1)
+            .processedWith(new AutoValueProcessor())
+            .compilesWithoutError()
+            .and().generatesSources(expected);
+  }
+
+  @Test public void multipleGenerics () {
+    JavaFileObject source1 = JavaFileObjects.forSourceString("test.Foo", ""
+            + "package test;\n"
+            + "import com.google.auto.value.AutoValue;\n"
+            + "import com.squareup.moshi.JsonAdapter;\n"
+            + "import com.squareup.moshi.Moshi;\n"
+            + "import java.lang.reflect.Type;\n"
+            + "import java.util.List;\n"
+            + "@AutoValue public abstract class Foo<V, T> {\n"
+            + "  public static <V, T>JsonAdapter<Foo<V, T>> jsonAdapter(Moshi moshi, Type[] types) {\n"
+            + "    return null;"
+            + "  }\n"
+            + "  public abstract List<V> items();\n"
```

Comment:
Since we've run into this issue before in other extensions, it would be helpful to have a test with types that aren't in order.  i.e. 
```Foo<V, T> {
  public abstract List<T> items();
  public abstract V foo();
}```